storage/DevicegraphImpl.h

```diff
@@ -111,58 +111,123 @@ namespace storage
 	vector<vertex_descriptor> leaves(vertex_descriptor vertex, bool itself) const;
 	vector<vertex_descriptor> roots(vertex_descriptor vertex, bool itself) const;
 
-	template <typename Type>
+
+	template<typename Type>
 	vector<Type*>
-	getDevices(const vector<vertex_descriptor>& vertices) // TODO rename
+	get_devices_of_type() const
 	{
 	    vector<Type*> ret;
 
-	    for (vertex_descriptor vertex : vertices)
+	    for (vertex_descriptor vertex : vertices())
 	    {
 		Type* device = dynamic_cast<Type*>(graph[vertex].get());
-		if (!device)
-		    throw std::bad_cast();
+		if (device)
+		    ret.push_back(device);
+	    }
+
+	    return ret;
+	}
+
+
+	template <typename Type, class StrictWeakOrdering>
+	vector<Type*>
+	get_devices_of_type(StrictWeakOrdering compare) const
+	{
+	    vector<Type*> ret = get_devices_of_type<Type>();
+
+	    sort(ret.begin(), ret.end(), compare);
+
+	    return ret;
+	}
+
 
-		ret.push_back(device);
+	template <typename Type, typename Pred>
```

Comment:
That would make all uses to the not _if functions more complicated (passing extra parameter). Since the predicate is not the last parameter a default value is not possible.